src/mo_frontend/test_recovery.ml

```diff
@@ -256,9 +397,102 @@ actor Main {
+      } (e.g. '}')
+      .<nat> (e.g. '.<nat>')
+      ! (e.g. '!')
+      <exp_nullary(ob)> (e.g. '42')
+      <unop> <exp_bin(ob)> (e.g. '^ 42')
+      <unassign> <exp(ob)> (e.g. '^= 42')
+      <binop> <exp_bin(ob)> (e.g. '-% 42')
+      <binassign> <exp(ob)> (e.g. '-%= 42')
+      <binop> <exp_bin(ob)> (e.g. '**% 42')
+      <binassign> <exp(ob)> (e.g. '**%= 42')
+      <binop> <exp_bin(ob)> (e.g. '*% 42')
+      <binassign> <exp(ob)> (e.g. '*%= 42')
+      <binop> <exp_bin(ob)> (e.g. '+% 42')
+      <binassign> <exp(ob)> (e.g. '+%- 42')
+      <unop> <exp_bin(ob)> (e.g. '- 42')
+      <binop> <exp_bin(ob)> (e.g. ' >> 42')
+      <binop> <exp(ob)> (e.g. '>>= 42')
+      <binop> <exp_bin(ob)> (e.g. '<< 42')
+      <binassign> <exp(ob)> (e.g. '<<= 42')
+      ; seplist(<dec_field>,<semicolon>) (e.g. '; public let x : Int = 0')
+      <binop> <exp_bin(ob)> (e.g. '<>> 42')
+      <binassign> <exp(ob)> (e.g. '<>>= 42')
+      <binop> <exp_bin(ob)> (e.g. '<<> 42')
+      <binassign> <exp(ob)> (e.g. '<<>= 42')
+      <binop> <exp_bin(ob)> (e.g. '** 42')
+      <binassign> <exp(ob)> (e.g. '**= 42')
+      <unassign> <exp(ob)> (e.g. '+= 42')
+      |> <exp_bin(ob)> (e.g. '|> 42')
+      <binop> <exp_bin(ob)> (e.g. '| 42')
+      <binassign> <exp(ob)> (e.g. '|= 42')
+      or <exp_bin(ob)> (e.g. 'or 42')
+      <binop> <exp_bin(ob)> (e.g. '!= 42')
+      <binop> <exp_bin(ob)> (e.g. '* 42')
+      <binassign> <exp(ob)> (e.g. '*= 42')
+      <binop> <exp_bin(ob)> (e.g. '% 42')
+      <binassign> <exp(ob)> (e.g. '%= 42')
+      <unassign> <exp(ob)> (e.g. '-= 42')
+      <relop> <exp_bin(ob)> (e.g. ' <  42')
+      <relop> <exp_bin(ob)> (e.g. '<= 42')
+      implies <exp_bin(ob)> (e.g. 'implies 42')
+      <binop> <exp_bin(ob)> (e.g. '# 42')
+      <relop> <exp_bin(ob)> (e.g. ' >  42')
+      <relop> <exp_bin(ob)> (e.g. '>= 42')
+      <relop> <exp_bin(ob)> (e.g. '== 42')
+      else <exp_nest> (e.g. 'else 42')
+      . <id> (e.g. '. <id>')
+      <binop> <exp_bin(ob)> (e.g. '/ 42')
+      <binassign> <exp(ob)> (e.g. '/= 42')
+      : <typ_nobin> (e.g. ': Int')
+      <binassign> <exp(ob)> (e.g. '@= 42')
+      <binassign> <exp(ob)> (e.g. 'assign 42')
+      <binop> <exp_bin(ob)> (e.g. '& 42')
+      <binassign> <exp(ob)> (e.g. '&= 42')
+      and <exp_bin(ob)> (e.g. 'and 42')
+      <unop> <exp_bin(ob)> (e.g. '+ 42')
+      <inst> <exp_nullary(ob)> (e.g. '<Int> 42')
+      [ <exp(ob)> ] (e.g. '[ 42 ]')
+
+    (unknown location): syntax error [M0001], unexpected token '(', expected one of token or <phrase> sequence:
+      func <pat_plain> <annot_opt> <func_body> (e.g. 'func x : Int {}')
+      class <pat_plain> <annot_opt> <class_body> (e.g. 'class x : Int = {}')
+      object class <pat_plain> <annot_opt> <class_body> (e.g. 'object class x : Int = {}')
+      module class <pat_plain> <annot_opt> <class_body> (e.g. 'module class x : Int = {}')
```

Comment:
These examples are syntactically ok, but would never type checked because the pattern are untyped.
I wonder if we can improve this by introducing
<arg_pat> := <pat_plain>
in the grammar and using it at appropriat places,
and then using a different example for <arg_pat> (such as ( x : Int)).
Maybe too hacky.

Reply:
That seems to work nicely in #5108. Not saying you should merge that it was just an experiment.